docs/decisions/0054-python-structured-output.md

```diff
@@ -0,0 +1,312 @@
+
+To address this issue, OpenAI has introduced **Structured Outputs**—a feature designed to ensure that model-generated outputs conform precisely to developer-specified JSON Schemas. This advancement allows developers to build more robust applications by providing guarantees that AI outputs will match predefined structures, improving interoperability with downstream systems.
+
+In recent evaluations, the new GPT-4o-2024-08-06 model with Structured Outputs demonstrated a perfect 100% score in adhering to complex JSON schemas, compared to GPT-4-0613, which scored less than 40%. Structured Outputs streamline the process of generating reliable structured data from unstructured inputs, a core need in various AI-powered applications such as data extraction, automated workflows, and function calling.
+
+---
+
+## Problem Statement
+
+Developers building AI-driven solutions using the OpenAI API often face challenges when extracting structured data from unstructured inputs. Ensuring model outputs conform to predefined JSON schemas is critical for creating reliable and interoperable systems. However, current models, even with JSON mode, do not guarantee schema conformity, leading to inefficiencies, errors, and additional development overhead in the form of retries and custom tools.
+
+With the introduction of Structured Outputs, OpenAI models are now able to strictly adhere to developer-provided JSON schemas. This feature eliminates the need for cumbersome workarounds and provides a more streamlined, efficient way to ensure consistency and reliability in model outputs. Integrating Structured Outputs into the Semantic Kernel orchestration SDK will enable developers to create more powerful, schema-compliant applications, reduce errors, and improve overall productivity.
+
+## Using Structured Outputs
+
+### 1. Function Calling
+
+Structured outputs can be used for function calling to ensure adherence to a schema. In Python, using OpenAI's latest SDK, developers can use the `pydantic_function_tool()` method to convert a Pydantic model into the required `structured output` function-calling JSON schema:
+
+
+```python
+client = AsyncOpenAI()
+
+# Note that the SDK method is `client.beta.chat.completions.parse` instead of our current `client.chat.completions.create` call.
+completion = await client.beta.chat.completions.parse(
+    model="gpt-4o-2024-08-06",
+    messages=[
+        {
+            "role": "system",
+            "content": "You are a helpful assistant. The current date is August 6, 2024. You help users query for the data they are looking for by calling the query function.",
+        },
+        {
+            "role": "user",
+            "content": "look up all my orders in may of last year that were fulfilled but not delivered on time",
+        },
+    ],
+    tools=[
+        openai.pydantic_function_tool(SomeClass), # This is a new method provided by OpenAI to handle Pydantic models
```

Comment:
Does it load all functions from `SomeClass`?

Reply:
Let's say you have the following:

```python
class GetDeliveryDate(BaseModel):
    order_id: str

tools = [openai.pydantic_function_tool(GetDeliveryDate)]
```

That is making sure that there is a schema defined so the model knows about `GetDeliveryDate` and its attributes. For our purposes, I would prefer to maintain how we build the JSON Schema in KernelParameterMetadata, and then add the other required function calling structure, as needed by the model. Not everyone is going to use Pydantic, so we don't need to solely rely on `openai.pydantic_function_tool`.